src/numericalnim/utils.nim

```diff
@@ -262,104 +264,173 @@ proc calcError*[T](y_true, y: T): float {.inline.} = mean_squared_error(y_true,
 
 
 proc hermiteSpline*[T](x, x1, x2: float, y1, y2, dy1, dy2: T): T {.inline.}=
-    let t = (x - x1)/(x2 - x1)
-    let h00 = (1.0 + 2.0 * t) * (1.0 - t) ^ 2
-    let h10 = t * (1.0 - t) ^ 2
-    let h01 = t ^ 2 * (3.0 - 2.0 * t)
-    let h11 = t ^ 3 - t ^ 2
-    result = h00 * y1 + h10 * (x2 - x1) * dy1 + h01 * y2 + h11 * (x2 - x1) * dy2
+  let t = (x - x1)/(x2 - x1)
+  let h00 = (1.0 + 2.0 * t) * (1.0 - t) ^ 2
+  let h10 = t * (1.0 - t) ^ 2
+  let h01 = t ^ 2 * (3.0 - 2.0 * t)
+  let h11 = t ^ 3 - t ^ 2
+  result = h00 * y1 + h10 * (x2 - x1) * dy1 + h01 * y2 + h11 * (x2 - x1) * dy2
 
 
 proc hermiteInterpolate*[T](x: openArray[float], t: openArray[float],
-                            y, dy: openArray[T]): seq[T] {.inline.} =
-    # loop over each interval and check if x is in there, if x is sorted
-    var xIndex = 0
-    if isSorted(x):
+              y, dy: openArray[T]): seq[T] {.inline.} =
+  # loop over each interval and check if x is in there, if x is sorted
```

Comment:
I'm not sure if you know exactly how many elements will end up in the `result` sequence (probably not if only due to duplicate indices etc.). You could still add a `result = newSeqOfCap[T](…)` to at least reduce the number of reallocations (and possible have none).

Reply:
That is a great idea! :) I should document this proc better but I think `x` is the points I want to evaluate it at so `newSeqOfCap(x.len)` should work

Reply:
Worth noting as well is that this `hermiteInterpolate` is separate from the `HermiteSpline` in the `interpolate` module. Here we don't create an object, we just calculate the values directly.